include/swift/Basic/FileManager.h

```diff
@@ -0,0 +1,81 @@
+//===----------- FileManager.h - Status-caching file manager ----*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_BASIC_FILEMANAGER_H
+#define SWIFT_BASIC_FILEMANAGER_H
+
+#include "swift/Basic/LLVM.h"
+#include "clang/Basic/FileSystemStatCache.h"
+#include "llvm/Support/Errc.h"
+#include "llvm/Support/VirtualFileSystem.h"
+
+namespace llvm {
+namespace vfs {
+class FileSystem;
+}
+}
+
+namespace swift {
+
+/// A file manager that caches file stats, both to avoid re-statting files, but
+/// also to ensure we don't accidentally record newer stats than when we first
+/// read the file.
+class FileManager: public llvm::RefCountedBase<FileManager> {
+private:
+  friend class SourceManager;
+  clang::MemorizeStatCalls cachedStats;
+  llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> fileSystem;
+  StringRef getCachedFilename(StringRef path);
```

Comment:
The values returned here are invalidated if anyone calls `setFileSystem`.  It looks like this is used by the source manager.  Assuming this is actually okay in practice, we should document this pitfall.